enhancements/microshift/multus-cni-for-microshift.md

```diff
@@ -0,0 +1,445 @@
+                        ],
+                        "mac": "7a:ca:cf:19:64:e8",
+                        "dns": {},
+                        "gateway": [
+                            "\u003cnil\u003e"
+                        ]
+                    }]
+                  k8s.v1.cni.cncf.io/networks: bridge-conf,bridge-conf
+```
+
+There is an annotation for NetworkAttachmentDefinition - `k8s.v1.cni.cncf.io/resourceName`.
+It is used when some CNI requires information about specific device that is prepared by device plugin.
+It is primarily used with SR-IOV and as such out of scope for this enhancement mentioned only for completeness.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+First, it must be noted that Multus CNI itself is a meta-CNI. From a high level perspective, its
+purpose is to call other CNIs according to the CNI configs supplied by the user in form of
+NetworkAttachmentDefinitions and Pod annotations.
+Any specific actions are related to the delegate CNIs. For example: creating veth pair,
+attaching one end of veth to the bridge and making the other end available within the Pod is
+responsibility of `bridge` CNI is this example.
+MicroShift team will create tests for CNIs that will be declared as supported. However these tests
+will not explore the breadth and depth of possible network setups, so ultimately the responsibility
+for correctness of the configuration is up to the user.
+
+#### Manifests
+
+MicroShift will provide Multus based on thin architecture because of the resource consumption
+(see alternatives for more information).
+Because of the differences with OpenShift (which uses thick Multus plugin), existing OpenShift
+manifests will require changes to make them suitable for MicroShift. These manifests will reside
+in MicroShift repository.
+
+Updating necessary image references will be part of existing rebase procedure.
+
+#### RPM package
+
+RPM spec to build `microshift-multus` and `microshift-multus-release-info` RPMs will be part of
+existing `microshift.spec` file.
+The RPM will include:
+- manifests required to deploy Multus on MicroShift
+- CRI-O drop-in config to use Multus instead of ovn-kubernetes (which will require reorganization
+  of currently existing MicroShift's CRI-O configs)
+- greenboot healthcheck script
+- cleanup script plugin
+
+#### Container images: Multus and network plugins
+
+Because the Multus image used by OpenShift has a large size that is not acceptable for edge devices, new image
+will be prepared and will only include relevant artifacts such as entrypoint script/binary and
+Multus CNI binary (which is copied to host's `/opt/cni/bin`).
+
+To supply network plugins (CNIs) such as bridge, vlan, macvlan, and so on a new image will be prepared
+so MicroShift uses the same binaries as OpenShift (alternative is using RHEL's networkplugins RPM - see alternatives).
+This image will also contain IPAM binaries such as `static`, `dynamic (DHCP)`, and `host-local`.
```

Comment:
That's a good point, this makes sense. Let's ask @DanielFroehlich

Reply:
How many other things does the CNO do to manage Multus? Should we be considering building an operator based on a subset of that code?

Reply:
@zshi-redhat @s1061123 keep me honest but looking at the CNO's code, in regards to Multus, it's mostly rendering and deployment of [manifests](https://github.com/openshift/cluster-network-operator/tree/master/bindata/network/multus) and I see following pieces of templating logic: SDN/OVN, optional DHCP server, and optional whereabouts reconciler.

We learned that for dhcp/dynamic IPAM we need dhcp server. whereabouts is not required, but I think shipping it would mean just adding RBAC and DaemonSet, nothing fancy

Reply:
OK, that works for me. Please add that info to the alternatives section before you resolve this thread.

Reply:
whereabouts is not needed as microshift is single node, host-local shall be sufficient.
whereabouts is mainly to address the issue that host-local doesn't have a full view of IPs assigned to each node in a multi-node cluster, so it has a centralized controller that allocates/deallocates IPs cluster wide. host-local uses a local file to track which IP is assigned to which Pod, so it doesn't scale with two or more nodes.

For multus daemonset specifically, it renders multus plugin itself, containernetworking-plugins, whereabouts-cni, whereabouts-reconciler, bond-cni . Other multus related manifests in CNO includes multus-admission-controller, network-daemon-metrics, dhcp-daemonset.

multus-admission-controller, it was mentioned as non-goal in this enhancement
network-daemon-metrics, this is about exposing promethus metrics of how many net-attach-def is used and how many pods are using multus secondary network. not sure if this is of interest of microshift user.

Reply:
@zshi-redhat what's the difference between `dhcp` and `whereabouts`?
Sound like they're very similar, but whereabouts it complete packages whereas with `dhcp`, even though you have DHCP DaemonSet, you still need DHCP server?
So with dhcp you can assign IPs to Pods from your network, but how does it look with whereabouts?

Reply:
dhcp and whereabouts are two different IPAM plugins.
dhcp relies on external dhcp server to grant IPs, it requires dhcp-daemonset to restore the lease.
whereabouts allows user to define a range of IPs that can be allocated to multus secondary networks, it runs everything in the cluster itself and manages the IP pool to avoid overlapping and do garbage collection.

so both of them are used to assign IP to secondary pod network. for dhcp, pods get IP from external dhcp server. for whereabouts, pods get IP from whereabout controller.